management/WFCORE-5489_override_attribute_value_from_env_var.adoc

```diff
@@ -0,0 +1,152 @@
+
+== Resource and attribute mapping
+
+The mapping is straightforward. The name of the env var is based on thethe address of the resource and the name of the attribute:
+
+1. take the address of the resource
+** `/subsystem=undertow/server=default-server/http-listener=default`
+2. remove the leading slash (`/`)
+** `subsystem=undertow/server=default-server/http-listener=default`
+3. append *two underscores* (`__`) and the name of the attribute:
+** `subsystem=undertow/server=default-server/http-listener=default__proxy-address-forwarding`
+4. Replace all non-alphanumeric characters with an underscore (`_`) and put it in upper case
+** `SUBSYSTEM_UNDERTOW_SERVER_DEFAULT_SERVER_HTTP_LISTENER_DEFAULT__PROXY_ADDRESS_FORWARDING`
+
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFCORE-5489[As a developer, I want to override management attribute values using environment variables]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/WFCORE-5464[WFCORE-5464 - Add environment variables as a source for model expression resolution]
+** this is a different feature as it reads environment variables *if the attribute value is an expression*. With WFCORE-5489, we override the attribute value but do not specify its semantics (it can be an expression if the attribute allows it).
+
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Testing By
+
+* [x] Engineering
+
+=== Affected Projects or Components
+
+* https://github.com/wildfly/wildfly-core
+
+=== Other Interested Projects
+
+=== Relevant Installation Types
+
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+* [x] Managed domain
+* [x] OpenShift s2i
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+Everytime an attribute value is validated and set in the management model, we look if there is a corresponding environment value that overrides it *before* doing any expression resolution or correction.
+
+This change will be done in `org.jboss.as.controller.AttributeDefinition#validateAndSet`. Any attribute definition that does not use this method to validate and set their values will not be taken into account.
+
+This feature is activated by a presence of an environment variable: `WILDFLY_OVERRIDING_ENV_VARS`.
+
+As the target of this feature is application images running on the cloud, we will make it the default behaviour for the S2I images by setting this `WILDFLY_OVERRIDING_ENV_VARS` in the builder and runtime images.
+
+For cloud deployment with Bootable Jar, the users will have to set this env var in addtion to any env var that will overrides attribute values.
+
+=== Nice-to-Have Requirements
+
+=== Non-Requirements
+
+* This feature is not activated by default in other installation types than OpenShift S2I
+* It is not possible to override management attributes value using Java System Properties.
```

Comment:
* It is not a requirement that an attribute value being determined from an environment variable triggers persistence of the configuration in order to make its state match the current running configuration.
* Following from this it is not a requirement that files in the configuration history dir reflect environment variable values unless those history files record configuration state that was persisted for some other reason. For example, a standalone.xml.initial and standalone.xml.boot would not record information that wasn't in the config file that was read on the initial or most recent boot.

Reply:
added your comment verbatim :)